docs/extra/optimize_function.rst

```diff
@@ -0,0 +1,297 @@
+the best-fit parameters, the value at this location, a
+message, and a dictionary with optimizer-specific information:
+
+  >>> print(f"Success: {res[0]}")
+  Success: True
+  >>> print(f"Message: {res[3]}")
+  Message: successful termination
+  >>> print(f"extra:   {res[4]}")
+  extra:   {'info': 0, 'nfev': 88}
+  >>> print(f"best-fit location: {res[1]}")
+  best-fit location: [1.99999996 4.00064474]
+  >>> print(f"          minimum: {res[2]}")
+            minimum: 1.73832000295664e-11
+
+So, the correct location is (2, 4) and we can see we got close.
+
+As the different optimizers use the same interface we can easily try
+different optimizers:
+
+  >>> tbl = Table(names=['method', 'stat0', 'x', 'y'],
+  ...             dtype=[str, float, float, float])
+  >>> for method in [optfcts.minim, optfcts.neldermead, optfcts.lmdif, optfcts.montecarlo]:
+  ...     res = method(to_optimize, start, lo, hi)
+  ...     if res[0]:
+  ...         tbl.add_row([method.__name__, res[2], res[1][0], res[1][1]])
+  ...     else:
+  ...         print(f"Failed {method.__name__}: {res[3]}")
+  ...
+  Failed lmdif: improper input parameters
+  >>> print(tbl)
+      method           stat0                 x                  y
+    ---------- --------------------- ------------------ ------------------
+         minim  1.73832000295664e-11 1.9999999630648686  4.000644742396361
+    neldermead 4.932797434506576e-11 2.0000007686008545 3.9994204823962214
+    montecarlo 6.629393582604866e-12 1.9999997816942297  3.999564165982678
+
+In this particular case :py:func:`~sherpa.optmethods.optfcts.lmdif`
+failed, and this is because the ``to_optimize`` function returned None
+as its second argument. For the other cases we can see that they all
+found the minimum location.
+
+In order to use :py:func:`~sherpa.optmethods.optfcts.lmdif` we need
+the per-parameter statistic, which we can get with a small tweak::
+
+  def to_optimize2(args):
+      x = args[0]
+      y = args[1]
+      pa = (a - x) * (a - b)
+      pb = b * (y - x * x) * (y - x * x)
+      stat = pa * pa + pb * pb
+      return stat, [pa, pb]
+
+This lets us use `lmdif`::
+
+  >>> res2 = optfcts.lmdif(to_optimize2, start, lo, hi)
+  >>> res2[0]
+  True
+  >>> res2[1]
+  [2.        4.0000846]
+
+The `lmdif` optimizer is one of those that returns a number of
+elements in the "extra" dictionary::
+
+  >>> res2[4]
+  {'info': 0, 'nfev': 108, 'covar': array([[ 1.56250000e-02, -1.04361202e-01],
+         [-1.04361202e-01,  4.15098457e+03]]), 'num_parallel_map': 0}
+
+Optimizing a model
+------------------
+
+We can re-create the :ref:`Fitting the NormGauss1D and Lorentz1D models <fit_peaked_data_normgauss1d_lorentz1d>`
+section - at least for the :py:class:`~sherpa.models.basic.NormGauss1D` fit.
+
+The normalized gaussian model can be expressed as
+
+::
+
+  def ngauss(x, ampl, pos, fwhm):
+      term = 4 * np.log(2)
+      numerator = ampl * np.exp(-term * (x - pos) * (x - pos) / (fwhm * fwhm))
```

Comment:
Again `**2` to reduce the line length? This is long in my browser window, but of course it depends on the size of your screen and settings.